src/components/multi_select_state.js

```diff
@@ -109,41 +152,76 @@ const withMultiSelectState = WrappedComponent =>
             const index = items.findIndex(item => item.id === id);
             this.setState({ firstItemShiftSelected: index });
           }
+
           const newSelectedItems = this.getNewSelectedItems(id);
-          this.setState({ selectedItems: newSelectedItems }, this.handleChange);
+
+          const newFilteredSelectedItem = newSelectedItems.filter(
+            el => el.id == id
+          );
+
+          this.setState(
+            {
+              selectedItems: newSelectedItems,
+              filteredSelectedItems: [
+                ...filteredSelectedItems,
+                ...newFilteredSelectedItem
+              ]
+            },
+            this.handleChange
+          );
         }
       } else {
         this.unselectItems([id]);
       }
     }
 
     unselectItems(ids) {
-      const { selectedItems } = this.state;
+      const { selectedItems, filteredSelectedItems } = this.state;
+
       const newSelectedItems = selectedItems.filter(
         item => ids.find(id => id === item.id) === undefined
       );
+      const newFilteredSelectedItems = filteredSelectedItems.filter(
+        item => item.id !== ids[0]
+      );
+
       this.setState(
         {
-          selectedItems: newSelectedItems
+          selectedItems: newSelectedItems,
+          filteredSelectedItems: newFilteredSelectedItems
         },
         this.handleChange
       );
     }
 
     clearAll() {
-      this.setState({ selectedItems: [] }, this.handleChange);
+      this.setState(
+        { selectedItems: [], filteredSelectedItems: [] },
+        this.handleChange
+      );
     }
 
     filterItems(event) {
       const { items, filterFunction, searchValueChanged } = this.props;
       const { value } = event.target;
+
       this.setState({
         filteredItems: items.filter(filterFunction(value))
       });
 
       searchValueChanged && searchValueChanged(value);
     }
 
+    filterSelectedItems(event) {
+      const { filterFunction, searchSelectedItemsChanged } = this.props;
+      const { value } = event.target;
+      const { selectedItems } = this.state;
+
+      this.setState({
+        filteredSelectedItems: selectedItems.filter(filterFunction(value))
+      });
+      searchSelectedItemsChanged && searchSelectedItemsChanged(value);
+    }
     selectAllItems() {
```

Comment:
you didn't provide any changes to support select all logic
please check the following cases:

case 1:
1. select several items
2. add filter to right section
3. click select all
expected - in right section there should be displayed all items matching filter

case 2:
1. select all items
2. filter in right selection
3. click select all again
expected - right section should be empty